{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "09121ccb-941f-4904-8529-59d577d78214",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "''' \n",
    "AREA_output_statistics.ipynb \n",
    "By Natalie Bratset\n",
    "\n",
    "This notebook reads the AREA output file and generates a bunch of statistics based on that output. \n",
    "Make sure to change the export_folder variable to your exports folder!\n",
    "\n",
    "I'm planning to convert this to a .py file after writing/testing in this notebook.\n",
    "'''\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.patches as mplpatches\n",
    "import matplotlib.pyplot as plt\n",
    "# from scipy import stats\n",
    "# from scipy.interpolate import splrep, splev\n",
    "# import plotly.express as px\n",
    "\n",
    "area_results_path = os.path.expanduser(\"/scratch/Shares/dowell/temp/ChrisO/PSEA/AREA_fast/output/AREA_output/all_chrom_comorbidthreshold0/area_scores_20250808-192325.adjpval.csv\")  # Update with AREA rsults we want to look at\n",
    "gene_ref_path = os.path.expanduser(\"/scratch/Shares/dowell/temp/ChrisO/PSEA/AREA/testdata/genes.csv\")\n",
    "export_folder = \"/Users/nabr7467/projects/AREA_LE/AREA_output_statistics/\" # Change this to your export folder\n",
    "\n",
    "area_results = pd.read_csv(area_results_path)\n",
    "gene_ref = pd.read_csv(gene_ref_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "975a56e8-702b-4408-aa5f-9ee733edaaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "################################## USER INPUTS HERE ###########################################\n",
    "\n",
    "# this is for setting p_val cutoffs and whatnot\n",
    "# P value Columns: pval,p_value_bonf,p_value_holm,p_value_BenjaminiHochberg,p_value_BenjaminiYekutieli\n",
    "\n",
    "##############################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "14615a6c-2dc2-452d-8a61-4f17d1e4fef8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Down_syndrome plot completed\n",
      "Hashimoto_thyroiditis plot completed\n",
      "alopecia_areata plot completed\n",
      "anxiety plot completed\n",
      "asthma plot completed\n",
      "astigmatism plot completed\n",
      "atopic_eczema plot completed\n",
      "atrial_septal_defect plot completed\n",
      "atrioventricular_septal_defect plot completed\n",
      "attention_deficit.hyperactivity_disorder plot completed\n",
      "cardiac_rhythm_disease plot completed\n",
      "celiac_disease plot completed\n",
      "complete_trisomy_21 plot completed\n",
      "conductive_hearing_loss_disorder plot completed\n",
      "congenital_heart_disease plot completed\n",
      "constipation_disorder plot completed\n",
      "depressive_disorder plot completed\n",
      "digestive_system_disorder plot completed\n",
      "eustachian_tube_disorder plot completed\n",
      "eye_disorder plot completed\n",
      "folliculitis plot completed\n",
      "gastroesophageal_reflux_disease plot completed\n",
      "hearing_loss_disorder plot completed\n",
      "heart_disorder plot completed\n",
      "hidradenitis_suppurativa plot completed\n",
      "hypothyroidism plot completed\n",
      "myopia plot completed\n",
      "nervous_system_disorder plot completed\n",
      "obesity_disorder plot completed\n",
      "obstructive_sleep_apnea_syndrome plot completed\n",
      "otitis_media._susceptibility_to plot completed\n",
      "otorhinolaryngologic_disease plot completed\n",
      "patent_ductus_arteriosus plot completed\n",
      "patent_foramen_ovale plot completed\n",
      "pneumonia plot completed\n",
      "psychiatric_disorder plot completed\n",
      "pulmonary_hypertension plot completed\n",
      "reactive_airway_disease plot completed\n",
      "respiratory_system_disorder plot completed\n",
      "sensorineural_hearing_loss_disorder plot completed\n",
      "skeletal_system_disorder plot completed\n",
      "skin_disorder plot completed\n",
      "sleep_apnea_syndrome plot completed\n",
      "strabismus plot completed\n",
      "ventricular_septal_defect plot completed\n",
      "vitamin_D_deficiency plot completed\n"
     ]
    }
   ],
   "source": [
    "def add_gene_info(df, genes_df, gene_col_AREA, gene_col_ref):\n",
    "    '''Uses a gene reference dataframe to add gene info to the AREA data.'''\n",
    "    df1 = df.rename(columns={gene_col_AREA: gene_col_ref})\n",
    "    merged_df = pd.merge(df1, genes_df, on=gene_col_ref, how='left')\n",
    "    return merged_df\n",
    "\n",
    "def calc_linked_genes(df, group_col, pval_col, pval):\n",
    "    ''' doc '''\n",
    "    dictionary={}\n",
    "    df_grouped = df.groupby((group_col),observed=False)\n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        filtered_df = group[group[pval_col] < pval]\n",
    "        dictionary[name]=[len(filtered_df)]\n",
    "    genes = pd.DataFrame(dictionary)\n",
    "    \n",
    "    return genes\n",
    "\n",
    "def sum_comorbids():\n",
    "    pass\n",
    "\n",
    "def stats_export(df, genes_df, group_col, pval_col, pval):\n",
    "    data_info={}\n",
    "    \n",
    "    gene_list=df['value'].to_list()\n",
    "    gene_set=set(gene_list)\n",
    "    data_info['Total Genes Ran']=len(gene_set)\n",
    "    \n",
    "    all = calc_linked_genes(df, group_col, pval_col, pval) #this is a dataframe with # of genes with a pval < x for each comorbid\n",
    "    \n",
    "    ##filter DF by chr21 ##\n",
    "    chr21_genes = genes_df[genes_df['seqnames'] == '21']\n",
    "    chr21_df = df[df['value'].isin(chr21_genes['gene_id'].to_list())]\n",
    "    trisomy = calc_linked_genes(chr21_df, group_col, pval_col, pval) #this is a dataframe with # of genes on chr21 only with a pval < x for each comorbid\n",
    "    pass\n",
    "\n",
    "def get_num_zeros():\n",
    "    pass\n",
    "\n",
    "def minus_log_pval(df, col):\n",
    "    '''Calculates -log10 of specified pvalue column and exports the given dataframe with a new 'minuslog10pvalue' column.\n",
    "       This function handles pval=0 by finding the smallest non-zero p-value and treating all 0 p-values to the (smallest p-value) x (10^-10).\n",
    "       Inputs: df - AREA dataframe containing p-values \n",
    "              col - Column name with the pvalues you want to convert'''\n",
    "   \n",
    "    df['marker'] = '.' \n",
    "    \n",
    "    if 0 in area_results[col]:                                               # checks for a 0 pvalue to handle\n",
    "        df.loc[df[col] == 0.0, 'marker'] = 'v'\n",
    "        min_nonzero_pval = np.min(area_results[col][area_results[col] != 0]) # finds the minimum pvalue that is not 0\n",
    "        adjusted_pval = min_nonzero_pval/(10**10)                            # this value is the adjusted pvalue for AREA results with a pvalue of 0. This is such that we can take the -log10(pval)\n",
    "        df.loc[df[col] == 0, col] = adjusted_pval                            # change 0s to a smaller val than minimum\n",
    "    \n",
    "    df['minuslog10pvalue'] = -np.log10(df[col])\n",
    "    \n",
    "    return df\n",
    "\n",
    "def chr_manhattan_plot(df, group_col, pval_col, comorbid, export):\n",
    "    '''Creates a manhattan plot using a dataframe containing a column for chromosomes to split up, and plots the -log10(pval) column.'''\n",
    "    custom_order_dict = {'1' : 0,   '2': 1,   '3': 2,   '4': 3,\n",
    "                         '5' : 4,   '6': 5,   '7': 6,   '8': 7,\n",
    "                         '9' : 8,  '10': 9,  '11': 10, '12': 11,\n",
    "                         '13': 12, '14': 13, '15': 14, '16': 15,\n",
    "                         '17': 16, '18': 17, '19': 18, '20': 19,\n",
    "                         '21': 20, '22': 21,  'X': 22,  'Y': 23,\n",
    "                         'MT': 24}\n",
    "    \n",
    "    # df[group_col] = df[group_col].str.lower()           # makes everything lowercase for read-ablility\n",
    "    df_sorted = df.sort_values(by='seqnames',key=lambda x: x.map(custom_order_dict))\n",
    "    df_grouped = df_sorted.groupby((group_col), sort=False, observed=False) # groups dataframe by group_col (comorbidity)\n",
    "    \n",
    "    # Set Up Figure\n",
    "    figureHeight=8\n",
    "    figureWidth=14\n",
    "\n",
    "    plt.figure(figsize=(figureWidth,figureHeight),facecolor='white')\n",
    "\n",
    "    panelHeight=4\n",
    "    panelWidth=10\n",
    "\n",
    "    relativePanelWidth=panelWidth/figureWidth\n",
    "    relativePanelHeight=panelHeight/figureHeight\n",
    "\n",
    "    panel1=plt.axes([1/figureWidth,2/figureHeight,relativePanelWidth,relativePanelHeight])\n",
    "    \n",
    "    x_labels = []\n",
    "    x_labels_pos = []\n",
    "    \n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        x_labels.append(name)\n",
    "        x_labels_pos.append(num*2)\n",
    "        random_x_floats = np.random.uniform((num*2)-0.8,(num*2)+0.8,len(group[pval_col]))\n",
    "        group.reset_index(inplace=True)\n",
    "        panel1.scatter(random_x_floats, group[pval_col],marker='.',s=2.5)\n",
    "    # panel1.plot([-2,len(df_grouped)*2-0.8],[6,6],color='black', alpha=0.5)\n",
    "    panel1.set_xticks(x_labels_pos)\n",
    "    panel1.set_xticklabels(x_labels, rotation=45, ha='right')\n",
    "\n",
    "    panel1.set_yticks(np.arange(0,np.max(df[pval_col]),10))\n",
    "    \n",
    "    panel1.set_xlim(-1.2,len(df_grouped)*2-0.8)\n",
    "    panel1.set_ylim(-0.4,np.max(df[pval_col])+5)\n",
    "    \n",
    "    panel1.set_xlabel('All Chromosomes')\n",
    "    panel1.set_ylabel('- Log(p-value) for each gene')\n",
    "    \n",
    "    panel1.set_title(comorbid)\n",
    "    \n",
    "    plt.savefig(f'{export}manhattan_plot_{comorbid}_all_chr.png',dpi=600)\n",
    "    plt.close()\n",
    "    print(f'{comorbid} plot completed')\n",
    "\n",
    "    \n",
    "\n",
    "## MAIN ##\n",
    "combined_df = add_gene_info(area_results, gene_ref, 'value', 'gene_id')\n",
    "df = minus_log_pval(combined_df,'p_value_bonf')\n",
    "\n",
    "# stats_export(area_results, gene_ref, 'binary_attribute','p_value_bonf', 0.000001)\n",
    "# custom_manhattan_plot(df,'binary_attribute','minuslog10pvalue',[], export_folder) \n",
    "\n",
    "## Plots all comorbids and all chr ##\n",
    "df_grouped = df.groupby(('binary_attribute'),observed=False)\n",
    "for num, (name, group) in enumerate(df_grouped):\n",
    "    chr_manhattan_plot(group,'seqnames','minuslog10pvalue',name, export_folder) \n",
    "\n",
    "#statistics\n",
    "    # genes run per comorbidity\n",
    "    # genes linked to comorbid (@ specific pval)\n",
    "    # genes on 21 linked w/ comorbid (@ specific pval)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ebbe855-7c74-4093-b771-04d4621dcb39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vir_AREA_upgraded",
   "language": "python",
   "name": "vir_area_upgraded"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
