{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "09121ccb-941f-4904-8529-59d577d78214",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "''' \n",
    "AREA_output_statistics.ipynb \n",
    "By Natalie Bratset\n",
    "\n",
    "This notebook reads the AREA output file and generates a bunch of statistics based on that output. \n",
    "Make sure to change the export_folder variable to your exports folder!\n",
    "\n",
    "I'm planning to convert this to a .py file after writing/testing in this notebook.\n",
    "'''\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.patches as mplpatches\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# All Patients; All chr\n",
    "area_results_path = os.path.expanduser(\"/scratch/Shares/dowell/temp/ChrisO/PSEA/AREA_fast/output/AREA_output/all_chrom_comorbidthreshold0/area_scores_20250808-192325.adjpval.csv\")\n",
    "export_folder = \"/Users/nabr7467/projects/AREA_LE/AREA_output_statistics/all_patients_all_chr/\" # Change this to your export folder\n",
    "\n",
    "# All Patients; chr21 only\n",
    "# area_results_path = os.path.expanduser(\"/scratch/Shares/dowell/temp/ChrisO/PSEA/AREA_fast/output/AREA_output/chr21_comorbidthreshold0/area_scores_20250808-192418.adjpval.csv\")\n",
    "# export_folder = \"/Users/nabr7467/projects/AREA_LE/AREA_output_statistics/all_patients_chr21/\" # Change this to your export folder\n",
    "\n",
    "# T21 Patients; All chr\n",
    "# area_results_path = os.path.expanduser(\"/scratch/Shares/dowell/temp/ChrisO/PSEA/AREA_fast/output/AREA_output/completeT21_all_chrom/area_scores_20250812-120736.adjpval.csv\")\n",
    "# export_folder = os.path.expanduser(\"/Users/nabr7467/projects/AREA_LE/AREA_output_statistics/T21_patients_all_chr/\") # Change this to your export folder\n",
    "\n",
    "gene_ref_path = os.path.expanduser(\"/scratch/Shares/dowell/temp/ChrisO/PSEA/AREA/testdata/genes.csv\")\n",
    "# binary_attribute_path = os.path.expanduser(base_path + \"filtered_binary_attributes_dataframe.csv\")\n",
    "\n",
    "area_results = pd.read_csv(area_results_path)\n",
    "gene_ref = pd.read_csv(gene_ref_path)\n",
    "# binary_ref = pd.read_csv(binary_attribute_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "14615a6c-2dc2-452d-8a61-4f17d1e4fef8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        ENSG00000000003            ENSG00000000938       ENSG00000000971  \\\n",
      "0   complete_trisomy_21        complete_trisomy_21  patent_foramen_ovale   \n",
      "1                  None             hypothyroidism                  None   \n",
      "2                  None       sleep_apnea_syndrome                  None   \n",
      "3                  None  ventricular_septal_defect                  None   \n",
      "4                  None             celiac_disease                  None   \n",
      "5                  None                       None                  None   \n",
      "6                  None                       None                  None   \n",
      "7                  None                       None                  None   \n",
      "8                  None                       None                  None   \n",
      "9                  None                       None                  None   \n",
      "10                 None                       None                  None   \n",
      "11                 None                       None                  None   \n",
      "12                 None                       None                  None   \n",
      "13                 None                       None                  None   \n",
      "14                 None                       None                  None   \n",
      "15                 None                       None                  None   \n",
      "\n",
      "             ENSG00000001084            ENSG00000001461   ENSG00000001497  \\\n",
      "0   skeletal_system_disorder  ventricular_septal_defect  obesity_disorder   \n",
      "1                       None                       None              None   \n",
      "2                       None                       None              None   \n",
      "3                       None                       None              None   \n",
      "4                       None                       None              None   \n",
      "5                       None                       None              None   \n",
      "6                       None                       None              None   \n",
      "7                       None                       None              None   \n",
      "8                       None                       None              None   \n",
      "9                       None                       None              None   \n",
      "10                      None                       None              None   \n",
      "11                      None                       None              None   \n",
      "12                      None                       None              None   \n",
      "13                      None                       None              None   \n",
      "14                      None                       None              None   \n",
      "15                      None                       None              None   \n",
      "\n",
      "                ENSG00000001561      ENSG00000001631      ENSG00000002549  \\\n",
      "0      skeletal_system_disorder  complete_trisomy_21  complete_trisomy_21   \n",
      "1        pulmonary_hypertension       heart_disorder       hypothyroidism   \n",
      "2   respiratory_system_disorder                 None                 None   \n",
      "3                          None                 None                 None   \n",
      "4                          None                 None                 None   \n",
      "5                          None                 None                 None   \n",
      "6                          None                 None                 None   \n",
      "7                          None                 None                 None   \n",
      "8                          None                 None                 None   \n",
      "9                          None                 None                 None   \n",
      "10                         None                 None                 None   \n",
      "11                         None                 None                 None   \n",
      "12                         None                 None                 None   \n",
      "13                         None                 None                 None   \n",
      "14                         None                 None                 None   \n",
      "15                         None                 None                 None   \n",
      "\n",
      "        ENSG00000002586  ...      ENSG00000284606       ENSG00000284620  \\\n",
      "0   complete_trisomy_21  ...  complete_trisomy_21   complete_trisomy_21   \n",
      "1                  None  ...       hypothyroidism  sleep_apnea_syndrome   \n",
      "2                  None  ...                 None                  None   \n",
      "3                  None  ...                 None                  None   \n",
      "4                  None  ...                 None                  None   \n",
      "5                  None  ...                 None                  None   \n",
      "6                  None  ...                 None                  None   \n",
      "7                  None  ...                 None                  None   \n",
      "8                  None  ...                 None                  None   \n",
      "9                  None  ...                 None                  None   \n",
      "10                 None  ...                 None                  None   \n",
      "11                 None  ...                 None                  None   \n",
      "12                 None  ...                 None                  None   \n",
      "13                 None  ...                 None                  None   \n",
      "14                 None  ...                 None                  None   \n",
      "15                 None  ...                 None                  None   \n",
      "\n",
      "   ENSG00000284634      ENSG00000284657           ENSG00000284668  \\\n",
      "0   hypothyroidism  complete_trisomy_21  congenital_heart_disease   \n",
      "1             None                 None       complete_trisomy_21   \n",
      "2             None                 None                      None   \n",
      "3             None                 None                      None   \n",
      "4             None                 None                      None   \n",
      "5             None                 None                      None   \n",
      "6             None                 None                      None   \n",
      "7             None                 None                      None   \n",
      "8             None                 None                      None   \n",
      "9             None                 None                      None   \n",
      "10            None                 None                      None   \n",
      "11            None                 None                      None   \n",
      "12            None                 None                      None   \n",
      "13            None                 None                      None   \n",
      "14            None                 None                      None   \n",
      "15            None                 None                      None   \n",
      "\n",
      "                             ENSG00000284678        ENSG00000284681  \\\n",
      "0   attention_deficit.hyperactivity_disorder  hearing_loss_disorder   \n",
      "1                                       None                   None   \n",
      "2                                       None                   None   \n",
      "3                                       None                   None   \n",
      "4                                       None                   None   \n",
      "5                                       None                   None   \n",
      "6                                       None                   None   \n",
      "7                                       None                   None   \n",
      "8                                       None                   None   \n",
      "9                                       None                   None   \n",
      "10                                      None                   None   \n",
      "11                                      None                   None   \n",
      "12                                      None                   None   \n",
      "13                                      None                   None   \n",
      "14                                      None                   None   \n",
      "15                                      None                   None   \n",
      "\n",
      "        ENSG00000284713                   ENSG00000284716 ENSG00000284730  \n",
      "0   complete_trisomy_21                    heart_disorder         anxiety  \n",
      "1                  None  conductive_hearing_loss_disorder            None  \n",
      "2                  None                    hypothyroidism            None  \n",
      "3                  None                              None            None  \n",
      "4                  None                              None            None  \n",
      "5                  None                              None            None  \n",
      "6                  None                              None            None  \n",
      "7                  None                              None            None  \n",
      "8                  None                              None            None  \n",
      "9                  None                              None            None  \n",
      "10                 None                              None            None  \n",
      "11                 None                              None            None  \n",
      "12                 None                              None            None  \n",
      "13                 None                              None            None  \n",
      "14                 None                              None            None  \n",
      "15                 None                              None            None  \n",
      "\n",
      "[16 rows x 8937 columns]\n"
     ]
    }
   ],
   "source": [
    "def add_gene_info(df, genes_df, gene_col_AREA, gene_col_ref):\n",
    "    '''Uses a gene reference dataframe to add gene info to the AREA data.'''\n",
    "    df1 = df.rename(columns={gene_col_AREA: gene_col_ref})\n",
    "    merged_df = pd.merge(df1, genes_df, on=gene_col_ref, how='left')\n",
    "    return merged_df\n",
    "\n",
    "def calc_linked_genes(df, group_col, pval_col, pval):\n",
    "    ''' doc '''\n",
    "    dictionary={}\n",
    "    df_grouped = df.groupby((group_col),observed=False)\n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        filtered_df = group[group[pval_col] < pval]\n",
    "        dictionary[name]=[len(filtered_df)]\n",
    "    genes = pd.DataFrame(dictionary)\n",
    "    \n",
    "    return genes\n",
    "\n",
    "def calc_comorbid_per_gene(df, group_col, pval_col, pval):\n",
    "    ''' doc '''\n",
    "    dictionary_sum={}\n",
    "    dictionary_comorbids={}\n",
    "    df_grouped = df.groupby((group_col),observed=False)\n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        filtered_df = group[group[pval_col] < pval]\n",
    "        dictionary_sum[name]=[len(filtered_df)]\n",
    "        if len(filtered_df) != 0:\n",
    "            dictionary_comorbids[name] = filtered_df['binary_attribute'].to_list()\n",
    "            \n",
    "        \n",
    "    counts = pd.DataFrame(dictionary_sum)\n",
    "    \n",
    "    comorbids = pd.DataFrame.from_dict(dictionary_comorbids, orient='index')\n",
    "    comorbids = comorbids.transpose()\n",
    "    print(comorbids)\n",
    "    return counts, comorbids\n",
    "\n",
    "def sum_comorbids():\n",
    "    pass\n",
    "\n",
    "def stats_export(df, genes_df, group_col, pval_col, pval):\n",
    "    data_info={}\n",
    "    \n",
    "    # Total No. Genes Ran\n",
    "    gene_list=df['value'].to_list()\n",
    "    gene_set=set(gene_list)\n",
    "    data_info['Total Genes Ran']=len(gene_set)\n",
    "    \n",
    "    # No. Genes significant per comorbid\n",
    "    all = calc_linked_genes(df, group_col, pval_col, pval) #this is a dataframe with # of genes with a pval < x for each comorbid\n",
    "    data_info['Genes linked to each comorbid'] = all\n",
    "    \n",
    "    # No. Genes on chr21 significant per comorbid\n",
    "    chr21_genes = genes_df[genes_df['seqnames'] == '21']\n",
    "    chr21_df = df[df['value'].isin(chr21_genes['gene_id'].to_list())]\n",
    "    trisomy = calc_linked_genes(chr21_df, group_col, pval_col, pval) #this is a dataframe with # of genes on chr21 only with a pval < x for each comorbid\n",
    "    data_info['Genes on chr21 linked to each comorbid'] = trisomy\n",
    "    \n",
    "    # No. comorbids linked to each gene\n",
    "    counts, comorbids = calc_comorbid_per_gene(df, 'value' , pval_col, pval)\n",
    "    # print(df2)\n",
    "    \n",
    "\n",
    "def get_num_zeros():\n",
    "    pass\n",
    "\n",
    "def minus_log_pval(df, col):\n",
    "    '''Calculates -log10 of specified pvalue column and exports the given dataframe with a new 'minuslog10pvalue' column.\n",
    "       This function handles pval=0 by finding the smallest non-zero p-value and treating all 0 p-values to the (smallest p-value) x (10^-10).\n",
    "       Inputs: df - AREA dataframe containing p-values \n",
    "              col - Column name with the pvalues you want to convert'''\n",
    "   \n",
    "    df['marker'] = '.' \n",
    "    \n",
    "    if 0 in area_results[col]:                                               # checks for a 0 pvalue to handle\n",
    "        df.loc[df[col] == 0.0, 'marker'] = 'v'\n",
    "        min_nonzero_pval = np.min(area_results[col][area_results[col] != 0]) # finds the minimum pvalue that is not 0\n",
    "        adjusted_pval = min_nonzero_pval/(10**10)                            # this value is the adjusted pvalue for AREA results with a pvalue of 0. This is such that we can take the -log10(pval)\n",
    "        df[col+'_orig'] = df[col]                                            # Adds a col called col_orig to refer to\n",
    "        df.loc[df[col] == 0, col] = adjusted_pval                            # change 0s to a smaller val than minimum\n",
    "    \n",
    "    df['minuslog10pvalue'] = -np.log10(df[col])\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def comorbid_manhattan_plot(df, group_col, pval_col, export, patient_label):\n",
    "    '''Creates a manhattan plot for each comorbidity using a dataframe containing a column for groups to split up, and plots the -log10(pval) column.'''\n",
    "    df[group_col] = df[group_col].str.lower()           # makes everything lowercase for read-ablility\n",
    "    df = df.sort_values(group_col)\n",
    "    df_grouped = df.groupby((group_col),observed=False) # groups dataframe by group_col (comorbidity)\n",
    "    print(len(df_grouped))\n",
    "    \n",
    "    # Set Up Figure\n",
    "    figureHeight=8\n",
    "    figureWidth=14\n",
    "\n",
    "    plt.figure(figsize=(figureWidth,figureHeight),facecolor='white')\n",
    "\n",
    "    panelHeight=4\n",
    "    panelWidth=10\n",
    "\n",
    "    relativePanelWidth=panelWidth/figureWidth\n",
    "    relativePanelHeight=panelHeight/figureHeight\n",
    "\n",
    "    panel1=plt.axes([2/figureWidth,3/figureHeight,relativePanelWidth,relativePanelHeight])\n",
    "    \n",
    "    x_labels = []\n",
    "    x_labels_pos = []\n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        x_labels.append(name)\n",
    "        x_labels_pos.append(num*2)\n",
    "        random_x_floats = np.random.uniform((num*2)-0.8,(num*2)+0.8,len(group[pval_col]))\n",
    "        group.reset_index(inplace=True)\n",
    "        panel1.scatter(random_x_floats, group[pval_col],marker='.',s=2.5)\n",
    "    \n",
    "    # pval_cutoff = 0.0001\n",
    "    # panel1.plot([-2,len(df_grouped)*2-0.8],[- np.log10(pval_cutoff),- np.log10(pval_cutoff)],color='black', alpha=0.5) #line for significant pval\n",
    "    \n",
    "    panel1.set_xticks(x_labels_pos)\n",
    "    panel1.set_xticklabels(x_labels, rotation=45, ha='right')\n",
    "\n",
    "    panel1.set_yticks(np.arange(0,np.max(df[pval_col]),10))\n",
    "    \n",
    "    panel1.set_xlim(-1.2,len(df_grouped)*2-0.8)\n",
    "    panel1.set_ylim(-0.4,np.max(df[pval_col])+5)\n",
    "    \n",
    "    panel1.set_xlabel('Comorbidity')\n",
    "    panel1.set_ylabel('- Log(p-value)')\n",
    "    \n",
    "    panel1.set_title(f'AREA p-values of genes for {patient_label} Patients, by Comorbidity')\n",
    "    \n",
    "\n",
    "    if 0 in df['p_value_bonf_orig']:\n",
    "        adjusted_pval = np.min(df['p_value_bonf'])\n",
    "        plt.text(len(df_grouped)*2, np.max(df[pval_col]),'\\u2190',fontsize='large', va='center',ha='left')\n",
    "        plt.text(len(df_grouped)*2 + 2, np.max(df[pval_col]),f'p-value < {adjusted_pval:.1e}',fontsize='medium', va='center',ha='left')\n",
    "    \n",
    "    \n",
    "    plt.savefig(f'{export}manhattan_plot_{patient_label}_by_comorbid.png',dpi=600)\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    print('comorbid manhattan plot completed')\n",
    "\n",
    "    \n",
    "def chr_manhattan_plot(df, group_col, pval_col, comorbid, max_y, export, patient_label):\n",
    "    '''Creates a manhattan plot for each chromosome using a dataframe containing a column for chromosomes to split up, and plots the -log10(pval) column.'''\n",
    "    custom_order_dict = {'1' : 0,   '2': 1,   '3': 2,   '4': 3,\n",
    "                         '5' : 4,   '6': 5,   '7': 6,   '8': 7,\n",
    "                         '9' : 8,  '10': 9,  '11': 10, '12': 11,\n",
    "                         '13': 12, '14': 13, '15': 14, '16': 15,\n",
    "                         '17': 16, '18': 17, '19': 18, '20': 19,\n",
    "                         '21': 20, '22': 21,  'X': 22,  'Y': 23,\n",
    "                         'MT': 24}\n",
    "    \n",
    "    df_sorted = df.sort_values(by='seqnames',key=lambda x: x.map(custom_order_dict))\n",
    "    df_grouped = df_sorted.groupby((group_col), sort=False, observed=False) # groups dataframe by group_col (comorbidity)\n",
    "    \n",
    "    # Set Up Figure\n",
    "    figureHeight=6\n",
    "    figureWidth=14\n",
    "\n",
    "    plt.figure(figsize=(figureWidth,figureHeight),facecolor='white')\n",
    "\n",
    "    panelHeight=4\n",
    "    panelWidth=10\n",
    "\n",
    "    relativePanelWidth=panelWidth/figureWidth\n",
    "    relativePanelHeight=panelHeight/figureHeight\n",
    "\n",
    "    panel1=plt.axes([2/figureWidth,1/figureHeight,relativePanelWidth,relativePanelHeight])\n",
    "    \n",
    "    x_labels = []\n",
    "    x_labels_pos = []\n",
    "    \n",
    "    for num, (name, group) in enumerate(df_grouped):\n",
    "        x_labels.append(name)\n",
    "        x_labels_pos.append(num*2)\n",
    "        random_x_floats = np.random.uniform((num*2)-0.8,(num*2)+0.8,len(group[pval_col]))\n",
    "        group.reset_index(inplace=True)\n",
    "        panel1.scatter(random_x_floats, group[pval_col],marker='.',s=2.5)\n",
    "        \n",
    "    # panel1.plot([-2,len(df_grouped)*2-0.8],[6,6],color='grey', alpha=0.5) #line for significant pval\n",
    "    \n",
    "    panel1.set_xticks(x_labels_pos)\n",
    "    panel1.set_xticklabels(x_labels, rotation=45, ha='right')\n",
    "\n",
    "    panel1.set_yticks(np.arange(0,max_y,10))\n",
    "    \n",
    "    panel1.set_xlim(-1.2,len(df_grouped)*2-0.8)\n",
    "    panel1.set_ylim(-0.4,max_y+5)\n",
    "    \n",
    "    panel1.set_xlabel('All Chromosomes')\n",
    "    panel1.set_ylabel('- Log(p-value) for each gene')\n",
    "    \n",
    "    panel1.set_title(f'AREA p-values of genes for {patient_label} Patients with {comorbid}')\n",
    "    \n",
    "    if 0 in df['p_value_bonf_orig'].to_list(): ###how do i do this\n",
    "        adjusted_pval = np.min(df['p_value_bonf'])\n",
    "        plt.text(len(df_grouped)*2, np.max(df[pval_col]),'\\u2190',fontsize='large', va='center',ha='left')\n",
    "        plt.text(len(df_grouped)*2 + 1, np.max(df[pval_col]),f'p-value < {adjusted_pval:.1e}',fontsize='medium', va='center',ha='left')\n",
    "    \n",
    "    plt.savefig(f'{export}{patient_label}_{comorbid}_by_chr.png',dpi=600)\n",
    "    plt.close()\n",
    "    print(f'{comorbid} plot completed')\n",
    "\n",
    "    \n",
    "################################## USER INPUTS HERE ###########################################\n",
    "\n",
    "# this is for setting p_val cutoffs and whatnot\n",
    "# P value Columns: pval,p_value_bonf,p_value_holm,p_value_BenjaminiHochberg,p_value_BenjaminiYekutieli\n",
    "\n",
    "##############################################################################################\n",
    "\n",
    "## MAIN ##\n",
    "combined_df = add_gene_info(area_results, gene_ref, 'value', 'gene_id')\n",
    "df = minus_log_pval(combined_df,'p_value_bonf')\n",
    "max_pval = np.max(df['minuslog10pvalue'])\n",
    "\n",
    "stats_export(area_results, gene_ref, 'binary_attribute','p_value_bonf', 0.000001)\n",
    "\n",
    "## One Manhattan Plot, sorted by comorbidity ##\n",
    "# comorbid_manhattan_plot(df,'binary_attribute','minuslog10pvalue', export_folder, 'all')\n",
    "\n",
    "## Plots all comorbids and all chr ##\n",
    "# df_grouped = df.groupby(('binary_attribute'),observed=False)\n",
    "# for num, (name, group) in enumerate(df_grouped):\n",
    "#     chr_manhattan_plot(group,'seqnames','minuslog10pvalue',name, max_pval, export_folder, 'all') \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ebbe855-7c74-4093-b771-04d4621dcb39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vir_AREA_upgraded",
   "language": "python",
   "name": "vir_area_upgraded"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
